### PR TITLE
Implement general Mark Join

### PR DESCRIPTION
**Pull request by JackTan25**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #4153

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/sql/compile/compile.go [hunk 1]
```diff
@@ -694,9 +694,9 @@ func (c *Compile) compileJoin(n, right *plan.Node, ss []*Scope, children []*Scop
 		for i := range rs {
 			if isEq && len(conds) == 1 {
```
Comment (nnsgmsone) [inline pkg/sql/compile/compile.go:695]: Only mark join if len(conds)>1

**Review by nnsgmsone — CHANGES_REQUESTED** [review]

**Comment by nnsgmsone** [comment]
Realization of the existence of some problems.... Needs to be modified.

**Comment by nnsgmsone** [comment]
Please handle multiple columns

**Review by fengttt — COMMENTED** [review]
Test, test test.

All new operators, should have a bvt test that triggers this operator.    explain and confirm this operator is activated.

**Comment by JackTan25** [comment]
> Test, test test.
> 
> All new operators, should have a bvt test that triggers this operator. explain and confirm this operator is activated.

ok, I got it

**Comment by JackTan25** [comment]
> Test, test test.
> 
> All new operators, should have a bvt test that triggers this operator. explain and confirm this operator is activated.

I have add mark join, please wait Jensen to add multi col feature.

**Comment by JackTan25** [comment]
> Test, test test.
> 
> All new operators, should have a bvt test that triggers this operator. explain and confirm this operator is activated.

I‘ve added the mark-join bvt test. Now please wait jensen to finish the multiple columns feature.

**Review by fengttt — COMMENTED** [review]
Can you add a explain call on bvt test case and grep the result to see if it is really a mark join ...

**Comment by JackTan25** [comment]
> Can you add a explain call on bvt test case and grep the result to see if it is really a mark join ...

For now plan, if join result is marked as null, we don't output.

File: pkg/sql/colexec/mark/join.go [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+		var condNonEq resultType
+		var condEq resultType
+		var err error
+		for k := 0; k < n; k++ {
+			if ctr.inBuckets[k] == 0 {
+				continue
+			}
+			if zvals[k] == 0 { // 2.1 : probe tuple has null
+				condState = condFalse
+				for sel := 0; sel < ctr.bat.Length(); sel++ {
```
Comment (nnsgmsone) [inline pkg/sql/colexec/mark/join.go:224]: Please use the batch process.

File: pkg/sql/colexec/mark/join.go [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+		}
+	}
+}
+
+func (ctr *container) build(ap *Argument, proc *process.Process, anal process.Analyze) error {
+	bat := <-proc.Reg.MergeReceivers[1].Ch
+	if bat != nil {
+		joinMap := bat.Ht.(*hashmap.JoinMap)
+		ctr.evalNullSels(bat)
+		ctr.nullWithBatch = DumpBatch(bat, proc, ctr.nullSels)
```
Comment (nnsgmsone) [inline pkg/sql/colexec/mark/join.go:150]: These two copies don't seem to make any sense, please delete them

File: pkg/sql/compile/compile.go [hunk 4]
```diff
@@ -700,13 +700,12 @@ func (c *Compile) compileJoin(n, right *plan.Node, ss []*Scope, children []*Scop
 			}
 		}
 	case plan.Node_ANTI:
-		_, conds := extraJoinConditions(n.OnList)
 		for i := range rs {
-			if isEq && len(conds) == 1 {
+			if isEq {
```
Comment (nnsgmsone) [inline pkg/sql/compile/compile.go:704]: if len(conds) == 1, use anti join, only if len(conds) > 1, use mark join

**Review by nnsgmsone — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Pull request merged**